internal/x/xkube/provider.go

```diff
@@ -0,0 +1,251 @@
+			return &ServiceAccount{
+				SAToken:   string(sa.Data[saTokenKey]),
+				ClusterCA: string(sa.Data[saCertKey]),
+			}, nil
+		}
+	}
+
+	return nil, xerrors.Errorf("kubernetes service account secret with annotaion %s=%s not found", saNameAnnotationKey, ResourceName)
+}
+
+// InstallWorkspaceProviderResources creates the resources for a workspace provider.
+func InstallWorkspaceProviderResources(ctx context.Context, clientset *kubernetes.Clientset, namespace string) error {
+	err := installServiceAccount(ctx, clientset, namespace)
+	if err != nil {
+		return xerrors.Errorf("installing kubernetes service account: %w", err)
+	}
+
+	err = installWorkspaceServiceAccount(ctx, clientset, namespace)
+	if err != nil {
+		return xerrors.Errorf("installing workspace kubernetes service account: %w", err)
+	}
+
+	err = installRole(ctx, clientset, namespace)
+	if err != nil {
+		return xerrors.Errorf("installing kubernetes role: %w", err)
+	}
+
+	err = installRoleBinding(ctx, clientset, namespace)
+	if err != nil {
+		return xerrors.Errorf("installing kubernetes role binding: %w", err)
+	}
+
+	return nil
+}
+
+// PrettyRules returns a pretty format string of the coder role rules.
+func PrettyRules() string {
+	var output string
+	for _, rule := range roleSpec().Rules {
+		// core api group is represented as "" so we handle that formatting
+		for i, group := range rule.APIGroups {
+			if group == "" {
+				rule.APIGroups[i] = `""`
+			}
+		}
+		groups := strings.Join(rule.APIGroups, ",")
+		resources := strings.Join(rule.Resources, ",")
+		verbs := strings.Join(rule.Verbs, ",")
+		output = fmt.Sprintf("%s  - APIGroups: %s\n    Resources: %s\n    Verbs: %s\n", output, groups, resources, verbs)
+	}
+	return output
+}
+
+func installServiceAccount(ctx context.Context, clientset *kubernetes.Clientset, namespace string) error {
+	_, err := clientset.CoreV1().ServiceAccounts(namespace).Create(ctx, serviceAccountSpec(), metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			_, err = clientset.CoreV1().ServiceAccounts(namespace).Update(ctx, serviceAccountSpec(), metav1.UpdateOptions{})
+			if err != nil {
+				return xerrors.Errorf("updating service account: %w", err)
+			}
+			return nil
+		}
+		return xerrors.Errorf("creating service account: %w", err)
+	}
+
+	return nil
+}
+
+func installWorkspaceServiceAccount(ctx context.Context, clientset *kubernetes.Clientset, namespace string) error {
+	spec := serviceAccountSpec()
+	spec.Name = WorkspaceSA
+	_, err := clientset.CoreV1().ServiceAccounts(namespace).Create(ctx, spec, metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			_, err = clientset.CoreV1().ServiceAccounts(namespace).Update(ctx, spec, metav1.UpdateOptions{})
+			if err != nil {
+				return xerrors.Errorf("updating service account: %w", err)
+			}
+			return nil
+		}
+		return xerrors.Errorf("creating service account: %w", err)
+	}
+
+	return nil
+}
```

Comment:
could dedup these by making a helper function since they're mostly identical other than which spec it'll use